tsl/src/remote/dist_txn.c

```diff
@@ -474,11 +490,40 @@ dist_txn_xact_callback_2pc(XactEvent event, void *arg)
 static void
 dist_txn_xact_callback(XactEvent event, void *arg)
 {
+	bool use_2pc;
+	char *xactReadOnly;
+
 	/* Quick exit if no connections were touched in this transaction. */
 	if (store == NULL)
 		return;
 
-	if (ts_guc_enable_2pc)
+	/* Windows related shenanigans.. */
+	xactReadOnly = GetConfigOptionByName("transaction_read_only", NULL, false);
+
+	/*
+	 * The decision to use 2PC rests on multiple factors:
+	 *
+	 * 1) if ts_guc_enable_2pc is enabled and it's a regular backend use it
+	 *
+	 * 2) if ts_guc_enable_2pc is enabled but we are running a read only txn, don't use it
+	 *
+	 * 3) if (1) is true but only ONE DN is involved in the remote transaction, don't use it
+	 * For now, we hold off on (3) because it's possible that a transaction which involves
+	 * data on AN and DNs could get a failure at the fag end of the COMMIT processing on the
+	 * AN due to issues in local AN data. In such a case since we send a COMMIT at
+	 * "XACT_EVENT_PRE_COMMIT" event time to the DN, we might end up with a COMMITTED DN but
```

Comment:
Do you know why the commit is sent in the pre-commit hook as opposed to the commit hook? I guess it is because we need to be able to roll back the AN transaction if the commit on the DN fails, right? So, is (3) even a realistic option? Maybe we should just skip all mention of it here or, even better, clearly state that it is not a possibility if we want  to guarantee transactional semantics.

Reply:
Ok, removed this as a possibility.